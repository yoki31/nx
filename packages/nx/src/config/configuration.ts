import { readNxJson } from './nx-json';

/**
 * Returns information about where apps and libs will be created.
 */
export function workspaceLayout(): { appsDir: string; libsDir: string } {
  const nxJson = readNxJson();
  return {
    appsDir: nxJson.workspaceLayout?.appsDir ?? 'apps',
    libsDir: nxJson.workspaceLayout?.libsDir ?? 'libs',
  };
}

export { readNxJson } from './nx-json';
