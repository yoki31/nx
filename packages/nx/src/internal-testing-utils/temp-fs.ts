import { dirname, join } from 'path';
import { tmpdir } from 'os';
import {
  emptyDirSync,
  mkdirpSync,
  mkdtempSync,
  outputFile,
  readFile,
  realpathSync,
  rmSync,
  unlinkSync,
} from 'fs-extra';
import { joinPathFragments } from '../utils/path';
import { appendFileSync, existsSync, renameSync, writeFileSync } from 'fs';
import { setWorkspaceRoot, workspaceRoot } from '../utils/workspace-root';

type NestedFiles = {
  [fileName: string]: string;
};

export class TempFs {
  readonly tempDir: string;

  private previousWorkspaceRoot: string;

  constructor(private dirname: string, overrideWorkspaceRoot = true) {
    this.tempDir = realpathSync(mkdtempSync(join(tmpdir(), this.dirname)));
    this.previousWorkspaceRoot = workspaceRoot;

    if (overrideWorkspaceRoot) {
      setWorkspaceRoot(this.tempDir);
    }
  }

  async createFiles(fileObject: NestedFiles) {
    await Promise.all(
      Object.keys(fileObject).map(async (path) => {
        await this.createFile(path, fileObject[path]);
      })
    );
  }

  createFilesSync(fileObject: NestedFiles) {
    for (let path of Object.keys(fileObject)) {
      this.createFileSync(path, fileObject[path]);
    }
  }

  async createFile(filePath: string, content: string) {
    await outputFile(joinPathFragments(this.tempDir, filePath), content);
  }

  createFileSync(filePath: string, content: string) {
    let dir = joinPathFragments(this.tempDir, dirname(filePath));
    if (!existsSync(dir)) {
      mkdirpSync(dir);
    }
    writeFileSync(joinPathFragments(this.tempDir, filePath), content);
  }

  async readFile(filePath: string): Promise<string> {
    return await readFile(joinPathFragments(this.tempDir, filePath), 'utf-8');
  }

  removeFileSync(filePath: string): void {
    unlinkSync(joinPathFragments(this.tempDir, filePath));
  }

  appendFile(filePath: string, content: string) {
    appendFileSync(joinPathFragments(this.tempDir, filePath), content);
  }

  writeFile(filePath: string, content: string) {
    writeFileSync(joinPathFragments(this.tempDir, filePath), content);
  }
  renameFile(oldPath: string, newPath: string) {
    renameSync(
      joinPathFragments(this.tempDir, oldPath),
      joinPathFragments(this.tempDir, newPath)
    );
  }

  cleanup() {
    rmSync(this.tempDir, { recursive: true });
    setWorkspaceRoot(this.previousWorkspaceRoot);
  }

  reset() {
    emptyDirSync(this.tempDir);
  }
}
