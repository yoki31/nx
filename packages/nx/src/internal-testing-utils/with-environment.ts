export function withEnvironmentVariables<T>(
  env: Record<string, string | false | null | undefined>,
  callback: () => T
): T {
  const originalValues: Record<string, string> = {};
  for (const key in env) {
    originalValues[key] = process.env[key];
    const value = env[key];
    if (value) {
      process.env[key] = value;
    } else {
      delete process.env[key];
    }
  }
  const cleanup = () => {
    for (const key in env) {
      process.env[key] = originalValues[key];
    }
  };
  const p = callback();
  if (p instanceof Promise) {
    return p.finally(cleanup) as T;
  } else {
    cleanup();
    return p;
  }
}
