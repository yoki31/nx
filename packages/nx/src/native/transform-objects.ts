import { ProjectGraph } from '../config/project-graph';
import {
  ExternalNode,
  Project,
  Target,
  ProjectGraph as RustProjectGraph,
} from './index';

export function transformProjectGraphForRust(
  graph: ProjectGraph
): RustProjectGraph {
  const dependencies: Record<string, string[]> = {};
  const nodes: Record<string, Project> = {};
  const externalNodes: Record<string, ExternalNode> = {};
  for (const [projectName, projectNode] of Object.entries(graph.nodes)) {
    const targets: Record<string, Target> = {};
    for (const [targetName, targetConfig] of Object.entries(
      projectNode.data.targets ?? {}
    )) {
      targets[targetName] = {
        executor: targetConfig.executor,
        inputs: targetConfig.inputs,
        outputs: targetConfig.outputs,
        options: JSON.stringify(targetConfig.options),
        configurations: JSON.stringify(targetConfig.configurations),
      };
    }
    nodes[projectName] = {
      root: projectNode.data.root,
      namedInputs: projectNode.data.namedInputs,
      targets,
      tags: projectNode.data.tags,
    };
    if (graph.dependencies[projectName]) {
      dependencies[projectName] = [];
      for (const dep of graph.dependencies[projectName]) {
        dependencies[projectName].push(dep.target);
      }
    }
  }
  for (const [projectName, externalNode] of Object.entries(
    graph.externalNodes ?? {}
  )) {
    externalNodes[projectName] = {
      packageName: externalNode.data.packageName,
      hash: externalNode.data.hash,
      version: externalNode.data.version,
    };
    if (graph.dependencies[projectName]) {
      dependencies[projectName] = [];
      for (const dep of graph.dependencies[projectName]) {
        dependencies[projectName].push(dep.target);
      }
    }
  }

  return {
    nodes,
    externalNodes,
    dependencies,
  };
}
