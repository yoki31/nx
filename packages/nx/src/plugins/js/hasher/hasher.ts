import { ProjectGraphProjectNode } from '../../../config/project-graph';
import { readJsonFile } from '../../../utils/fileutils';
import { getRootTsConfigPath } from '../utils/typescript';
import {
  findProjectForPath,
  ProjectRootMappings,
} from '../../../project-graph/utils/find-project-for-path';

interface CompilerOptions {
  paths: Record<string, string[]>;
}

interface TsconfigJsonConfiguration {
  compilerOptions: CompilerOptions;
}

function readTsConfigJson(): TsconfigJsonConfiguration {
  try {
    const res = readJsonFile(getRootTsConfigPath());
    res.compilerOptions.paths ??= {};
    return res;
  } catch {
    return {
      compilerOptions: { paths: {} },
    };
  }
}

let tsConfigJson: TsconfigJsonConfiguration;

export function hashTsConfig(
  p: ProjectGraphProjectNode,
  projectRootMappings: ProjectRootMappings,
  { selectivelyHashTsConfig }: { selectivelyHashTsConfig: boolean }
) {
  if (!tsConfigJson) {
    tsConfigJson = readTsConfigJson();
  }
  if (selectivelyHashTsConfig) {
    return removeOtherProjectsPathRecords(p, tsConfigJson, projectRootMappings);
  } else {
    return JSON.stringify(tsConfigJson);
  }
}

function removeOtherProjectsPathRecords(
  p: ProjectGraphProjectNode,
  tsConfigJson: TsconfigJsonConfiguration,
  projectRootMapping: ProjectRootMappings
) {
  const { paths, ...compilerOptions } = tsConfigJson.compilerOptions;
  const filteredPaths: Record<string, string[]> = {};

  if (!paths) {
    return '';
  }

  for (const [key, files] of Object.entries(paths)) {
    for (const filePath of files) {
      if (p.name === findProjectForPath(filePath, projectRootMapping)) {
        filteredPaths[key] = files;
        break;
      }
    }
  }

  return JSON.stringify({
    compilerOptions: {
      ...compilerOptions,
      paths: filteredPaths,
    },
  });
}
