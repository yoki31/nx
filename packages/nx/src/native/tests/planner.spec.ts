import { TempFs } from '../../internal-testing-utils/temp-fs';
import { HashPlanner, transferProjectGraph } from '../index';
import { Task, TaskGraph } from '../../config/task-graph';
import { InProcessTaskHasher } from '../../hasher/task-hasher';
import { withEnvironmentVariables } from '../../internal-testing-utils/with-environment';
import { ProjectGraphBuilder } from '../../project-graph/project-graph-builder';
import { createTaskGraph } from '../../tasks-runner/create-task-graph';
import { transformProjectGraphForRust } from '../transform-objects';
import { DependencyType } from '../../config/project-graph';

let tempFs = new TempFs('task-planner');

describe('task planner', () => {
  // disable NX_NATIVE_TASK_HASHER for this test because we need to compare the results of the new planner with the old task hasher
  process.env.NX_NATIVE_TASK_HASHER = 'false';

  const packageJson = {
    name: 'nrwl',
  };

  const tsConfigBaseJson = JSON.stringify({
    compilerOptions: {
      paths: {
        '@nx/parent': ['libs/parent/src/index.ts'],
        '@nx/child': ['libs/child/src/index.ts'],
        '@nx/grandchild': ['libs/grandchild/src/index.ts'],
      },
    },
  });
  const allWorkspaceFiles = [
    { file: 'yarn.lock', hash: 'yarn.lock.hash' },
    { file: 'nx.json', hash: 'nx.json.hash' },
    { file: 'package-lock.json', hash: 'package-lock.json.hash' },
    { file: 'package.json', hash: 'package.json.hash' },
    { file: 'pnpm-lock.yaml', hash: 'pnpm-lock.yaml.hash' },
    { file: 'tsconfig.base.json', hash: tsConfigBaseJson },
    { file: 'workspace.json', hash: 'workspace.json.hash' },
    { file: 'global1', hash: 'global1.hash' },
    { file: 'global2', hash: 'global2.hash' },
  ];

  // TODO(cammisuli): This function is temporary until the new file hashing is implemented
  // This should just match snapshots of the planner
  async function assertHashPlan(
    task: Task | Task[],
    taskGraph: TaskGraph,
    taskHasher: InProcessTaskHasher,
    hashPlanner: HashPlanner
  ) {
    if (!Array.isArray(task)) task = [task];

    function getHashPlans(
      tasks: Task[],
      taskGraph: TaskGraph
    ): Record<string, string[]> {
      const plans = hashPlanner.getPlans(
        tasks.map((task) => task.id),
        taskGraph
      );

      for (const planId of Object.keys(plans)) {
        plans[planId] = plans[planId].sort();
      }

      return plans;
    }

    const hashes = await taskHasher.hashTasks(task, taskGraph);
    const plans = getHashPlans(task, taskGraph);

    // hashNodes here are completed in order because this is run with javascript.
    // we can then map the task id's by their index number with the hash nodes
    let hashNodes: Record<string, object> = task.reduce((acc, task, index) => {
      acc[task.id] = Object.keys(hashes[index].details.nodes).sort();
      return acc;
    }, {});

    for (let taskId of task.map((task) => task.id)) {
      expect(plans[taskId]).toEqual(hashNodes[taskId]);
    }

    return plans;
  }

  beforeEach(async () => {
    await tempFs.createFiles({
      'tsconfig.base.json': tsConfigBaseJson,
      'yarn.lock': 'content',
      'package.json': JSON.stringify(packageJson),
    });
  });

  afterEach(() => {
    tempFs.reset();
  });

  it('should build a plan that matches the original task-hasher', async () => {
    await withEnvironmentVariables({ TESTENV: 'env123' }, async () => {
      let projectFileMap = {
        parent: [{ file: '/file', hash: 'file.hash' }],
        unrelated: [{ file: 'libs/unrelated/filec.ts', hash: 'filec.hash' }],
      };

      const builder = new ProjectGraphBuilder();

      builder.addNode({
        name: 'parent',
        type: 'lib',
        data: {
          root: 'parent',
          targets: {
            build: {
              executor: 'nx:run-commands',
              inputs: [
                'default',
                '^default',
                { runtime: 'echo runtime123' },
                { env: 'TESTENV' },
                { env: 'NONEXISTENTENV' },
                {
                  input: 'default',
                  projects: ['unrelated', 'tag:some-tag'],
                },
              ],
            },
          },
        },
      });
      builder.addNode({
        name: 'unrelated',
        type: 'lib',
        data: {
          root: 'libs/unrelated',
          targets: { build: {} },
        },
      });
      builder.addNode({
        name: 'tagged',
        type: 'lib',
        data: {
          root: 'libs/tagged',
          targets: { build: {} },
          tags: ['some-tag'],
        },
      });
      const projectGraph = builder.getUpdatedProjectGraph();
      const taskGraph = createTaskGraph(
        projectGraph,
        {},
        ['parent'],
        ['build'],
        undefined,
        {},
        false
      );

      let nxJson = {} as any;

      const hasher = new InProcessTaskHasher(
        projectFileMap,
        allWorkspaceFiles,
        projectGraph,
        nxJson,
        null,
        {}
      );

      const ref = transferProjectGraph(
        transformProjectGraphForRust(projectGraph)
      );
      const planner = new HashPlanner(nxJson as any, ref);

      await assertHashPlan(
        taskGraph.tasks['parent:build'],
        taskGraph,
        hasher,
        planner
      );
    });
  });

  it('should plan the task where the project has dependencies', async () => {
    const projectFileMap = {
      parent: [
        { file: '/filea.ts', hash: 'a.hash' },
        { file: '/filea.spec.ts', hash: 'a.spec.hash' },
      ],
      child: [
        { file: '/fileb.ts', hash: 'b.hash' },
        { file: '/fileb.spec.ts', hash: 'b.spec.hash' },
      ],
      grandchild: [
        { file: '/filec.ts', hash: 'c.hash' },
        { file: '/filec.spec.ts', hash: 'c.spec.hash' },
      ],
    };

    const builder = new ProjectGraphBuilder(undefined, projectFileMap);

    builder.addNode({
      name: 'parent',
      type: 'lib',
      data: {
        root: 'libs/parent',
        targets: { build: { executor: 'unknown' } },
      },
    });

    builder.addNode({
      name: 'child',
      type: 'lib',
      data: {
        root: 'libs/child',
        targets: { build: { executor: 'none' } },
      },
    });
    builder.addNode({
      name: 'grandchild',
      type: 'lib',
      data: {
        root: 'libs/grandchild',
        targets: { build: { executor: 'none' } },
      },
    });
    builder.addStaticDependency('parent', 'child', '/filea.ts');
    builder.addStaticDependency('child', 'grandchild', '/fileb.ts');

    const projectGraph = builder.getUpdatedProjectGraph();

    let taskGraph = createTaskGraph(
      projectGraph,
      { build: ['^build'] },
      ['parent'],
      ['build'],
      undefined,
      {}
    );
    let nxJson = {} as any;
    const hasher = new InProcessTaskHasher(
      projectFileMap,
      allWorkspaceFiles,
      projectGraph,
      nxJson,
      null,
      {}
    );
    const planner = new HashPlanner(
      nxJson as any,
      transferProjectGraph(transformProjectGraphForRust(projectGraph))
    );
    const hashPlan = await assertHashPlan(
      taskGraph.tasks['parent:build'],
      taskGraph,
      hasher,
      planner
    );

    expect(hashPlan).toMatchSnapshot();
  });

  it('should plan non-default filesets', async () => {
    let projectFileMap = {
      parent: [
        { file: 'libs/parent/filea.ts', hash: 'a.hash' },
        { file: 'libs/parent/filea.spec.ts', hash: 'a.spec.hash' },
      ],
      child: [
        { file: 'libs/child/fileb.ts', hash: 'b.hash' },
        { file: 'libs/child/fileb.spec.ts', hash: 'b.spec.hash' },
      ],
    };

    let builder = new ProjectGraphBuilder(undefined, projectFileMap);

    builder.addNode({
      name: 'parent',
      type: 'lib',
      data: {
        root: 'libs/parent',
        targets: {
          build: {
            inputs: ['prod', '^prod'],
            executor: 'nx:run-commands',
          },
        },
      },
    });
    builder.addNode({
      name: 'child',
      type: 'lib',
      data: {
        root: 'libs/child',
        namedInputs: {
          prod: ['default'],
        },
        targets: { build: { executor: 'unknown' } },
      },
    });
    builder.addStaticDependency('parent', 'child', 'libs/parent/filea.ts');

    let projectGraph = builder.getUpdatedProjectGraph();

    let taskGraph = createTaskGraph(
      projectGraph,
      { build: ['^build'] },
      ['parent'],
      ['build'],
      undefined,
      {}
    );
    let nxJson = {
      namedInputs: {
        prod: ['!{projectRoot}/**/*.spec.ts'],
      },
    } as any;
    const hasher = new InProcessTaskHasher(
      projectFileMap,
      allWorkspaceFiles,
      projectGraph,
      nxJson,
      null,
      {}
    );
    const planner = new HashPlanner(
      nxJson as any,
      transferProjectGraph(transformProjectGraphForRust(projectGraph))
    );
    let hashPlans = await assertHashPlan(
      taskGraph.tasks['parent:build'],
      taskGraph,
      hasher,
      planner
    );

    expect(hashPlans).toMatchSnapshot();
  });

  it('should make a plan with multiple filesets of a project', async () => {
    let projectFileMap = {
      parent: [
        { file: 'libs/parent/filea.ts', hash: 'a.hash' },
        { file: 'libs/parent/filea.spec.ts', hash: 'a.spec.hash' },
      ],
    };
    let builder = new ProjectGraphBuilder(undefined, projectFileMap);
    builder.addNode({
      name: 'parent',
      type: 'lib',
      data: {
        root: 'libs/parent',
        targets: {
          build: {
            inputs: ['prod'],
            executor: 'nx:run-commands',
          },
          test: {
            inputs: ['default'],
            dependsOn: ['build'],
            executor: 'nx:run-commands',
          },
        },
      },
    });
    let projectGraph = builder.getUpdatedProjectGraph();

    let taskGraph = createTaskGraph(
      projectGraph,
      {},
      ['parent'],
      ['build', 'test'],
      undefined,
      {}
    );
    let nxJson = {
      namedInputs: {
        prod: ['!{projectRoot}/**/*.spec.ts'],
      },
    } as any;
    const hasher = new InProcessTaskHasher(
      projectFileMap,
      allWorkspaceFiles,
      projectGraph,
      nxJson,
      null,
      {}
    );
    const planner = new HashPlanner(
      nxJson as any,
      transferProjectGraph(transformProjectGraphForRust(projectGraph))
    );
    const tasks = Object.values(taskGraph.tasks);

    let plans = await assertHashPlan(tasks, taskGraph, hasher, planner);
    expect(plans).toMatchSnapshot();
  });

  it('should be able to handle multiple filesets per project', async () => {
    await withEnvironmentVariables(
      { MY_TEST_HASH_ENV: 'MY_TEST_HASH_ENV_VALUE' },
      async () => {
        let projectFileMap = {
          parent: [
            { file: 'libs/parent/filea.ts', hash: 'a.hash' },
            { file: 'libs/parent/filea.spec.ts', hash: 'a.spec.hash' },
          ],
          child: [
            { file: 'libs/child/fileb.ts', hash: 'b.hash' },
            { file: 'libs/child/fileb.spec.ts', hash: 'b.spec.hash' },
          ],
        };
        const builder = new ProjectGraphBuilder(undefined, projectFileMap);
        builder.addNode({
          name: 'parent',
          type: 'lib',
          data: {
            root: 'libs/parent',
            targets: {
              test: {
                inputs: ['default', '^prod'],
                executor: 'nx:run-commands',
              },
            },
          },
        });
        builder.addNode({
          name: 'child',
          type: 'lib',
          data: {
            root: 'libs/child',
            namedInputs: {
              prod: [
                '!{projectRoot}/**/*.spec.ts',
                '{workspaceRoot}/global2',
                { env: 'MY_TEST_HASH_ENV' },
              ],
            },
            targets: {
              test: {
                inputs: ['default'],
                executor: 'nx:run-commands',
              },
            },
          },
        });
        builder.addStaticDependency('parent', 'child', 'libs/parent/filea.ts');
        let projectGraph = builder.getUpdatedProjectGraph();
        let taskGraph = createTaskGraph(
          projectGraph,
          { build: ['^build'] },
          ['parent'],
          ['test'],
          undefined,
          {}
        );
        let nxJson = {
          namedInputs: {
            default: ['{projectRoot}/**/*', '{workspaceRoot}/global1'],
            prod: ['!{projectRoot}/**/*.spec.ts'],
          },
        };
        const hasher = new InProcessTaskHasher(
          projectFileMap,
          allWorkspaceFiles,
          projectGraph,
          nxJson as any,
          null,
          {}
        );

        const planner = new HashPlanner(
          nxJson as any,
          transferProjectGraph(transformProjectGraphForRust(projectGraph))
        );
        const tasks = Object.values(taskGraph.tasks);
        let plans = await assertHashPlan(tasks, taskGraph, hasher, planner);
        expect(plans).toMatchSnapshot();
      }
    );
  });

  it('should hash executors', async () => {
    let projectFileMap = {
      parent: [],
      child: [],
    };
    const builder = new ProjectGraphBuilder(undefined, projectFileMap);
    builder.addNode({
      name: 'proj',
      type: 'lib',
      data: {
        root: 'libs/proj',
        targets: {
          lint: {
            inputs: ['default'],
            executor: '@nx/eslint:lint',
          },
        },
      },
    });
    builder.addExternalNode({
      type: 'npm',
      name: 'npm:@nx/eslint',
      data: {
        packageName: '@nx/eslint',
        hash: 'hash1',
        version: '1.0.0',
      },
    });
    builder.addExternalNode({
      type: 'npm',
      name: 'npm:@nx/devkit',
      data: {
        packageName: '@nx/devkit',
        hash: 'hash2',
        version: '1.0.0',
      },
    });
    builder.addDependency(
      'npm:@nx/eslint',
      'npm:@nx/devkit',
      DependencyType.static
    );
    let projectGraph = builder.getUpdatedProjectGraph();
    let taskGraph = createTaskGraph(
      projectGraph,
      { build: ['^build'] },
      ['proj'],
      ['lint'],
      undefined,
      {}
    );
    let nxJson = {
      namedInputs: {
        default: ['{projectRoot}/**/*', '{workspaceRoot}/global1'],
        prod: ['!{projectRoot}/**/*.spec.ts'],
      },
    };
    const hasher = new InProcessTaskHasher(
      projectFileMap,
      allWorkspaceFiles,
      projectGraph,
      nxJson as any,
      null,
      {}
    );

    const planner = new HashPlanner(
      nxJson as any,
      transferProjectGraph(transformProjectGraphForRust(projectGraph))
    );
    const tasks = Object.values(taskGraph.tasks);
    let plans = await assertHashPlan(tasks, taskGraph, hasher, planner);
    expect(plans).toMatchSnapshot();
  });

  it('should build plans where the project graph has circular dependencies', async () => {
    let projectFileMap = {
      parent: [{ file: '/filea.ts', hash: 'a.hash' }],
      child: [{ file: '/fileb.ts', hash: 'b.hash' }],
    };
    let builder = new ProjectGraphBuilder(undefined, projectFileMap);
    builder.addNode({
      name: 'parent',
      type: 'lib',
      data: {
        root: 'libs/parent',
        targets: { build: { executor: 'nx:run-commands' } },
      },
    });
    builder.addNode({
      name: 'child',
      type: 'lib',
      data: {
        root: 'libs/child',
        targets: { build: { executor: 'nx:run-commands' } },
      },
    });
    builder.addStaticDependency('parent', 'child', '/filea.ts');
    builder.addStaticDependency('child', 'parent', '/fileb.ts');
    let projectGraph = builder.getUpdatedProjectGraph();
    let taskGraph = createTaskGraph(
      projectGraph,
      { build: ['^build'] },
      ['parent'],
      ['build'],
      undefined,
      {}
    );
    let nxJson = {} as any;
    const hasher = new InProcessTaskHasher(
      projectFileMap,
      allWorkspaceFiles,
      projectGraph,
      nxJson,
      null,
      {}
    );
    const planner = new HashPlanner(
      nxJson as any,
      transferProjectGraph(transformProjectGraphForRust(projectGraph))
    );
    let tasks = Object.values(taskGraph.tasks);
    let plans = await assertHashPlan(tasks, taskGraph, hasher, planner);
    expect(plans).toMatchSnapshot();
  });

  it('should include npm projects', async () => {
    let projectFileMap = {
      app: [{ file: '/filea.ts', hash: 'a.hash' }],
    };
    let builder = new ProjectGraphBuilder(undefined, projectFileMap);
    builder.addNode({
      name: 'app',
      type: 'app',
      data: {
        root: 'apps/app',
        targets: { build: { executor: 'nx:run-commands' } },
      },
    });
    builder.addExternalNode({
      name: 'npm:react',
      type: 'npm',
      data: {
        version: '17.0.0',
        packageName: 'react',
      },
    });
    builder.addStaticDependency('app', 'npm:react', '/filea.ts');
    let projectGraph = builder.getUpdatedProjectGraph();
    let taskGraph = createTaskGraph(
      projectGraph,
      { build: ['^build'] },
      ['app'],
      ['build'],
      undefined,
      {}
    );
    let nxJson = {} as any;
    const hasher = new InProcessTaskHasher(
      projectFileMap,
      allWorkspaceFiles,
      projectGraph,
      nxJson,
      null,
      {}
    );
    const transformed = transferProjectGraph(
      transformProjectGraphForRust(projectGraph)
    );
    const planner = new HashPlanner(nxJson as any, transformed);
    let plans = await assertHashPlan(
      taskGraph.tasks['app:build'],
      taskGraph,
      hasher,
      planner
    );
    expect(plans).toMatchSnapshot();
  });

  describe('dependentTasksOutputFiles', () => {
    it('should depend on dependent tasks output files', async () => {
      const projectFileMap = {
        parent: [
          { file: 'libs/parent/filea.ts', hash: 'a.hash' },
          { file: 'libs/parent/filea.spec.ts', hash: 'a.spec.hash' },
        ],
        child: [
          { file: 'libs/child/fileb.ts', hash: 'b.hash' },
          { file: 'libs/child/fileb.spec.ts', hash: 'b.spec.hash' },
        ],
        grandchild: [
          { file: 'libs/grandchild/filec.ts', hash: 'c.hash' },
          { file: 'libs/grandchild/filec.spec.ts', hash: 'c.spec.hash' },
        ],
      };

      let builder = new ProjectGraphBuilder(undefined, projectFileMap);
      builder.addNode({
        name: 'parent',
        type: 'lib',
        data: {
          root: 'libs/parent',
          targets: {
            build: {
              dependsOn: ['^build'],
              inputs: ['prod', 'deps'],
              executor: 'nx:run-commands',
              outputs: ['{workspaceRoot}/dist/{projectRoot}'],
            },
          },
        },
      });
      builder.addNode({
        name: 'child',
        type: 'lib',
        data: {
          root: 'libs/child',
          targets: {
            build: {
              dependsOn: ['^build'],
              inputs: ['prod', 'deps'],
              executor: 'nx:run-commands',
              outputs: ['{workspaceRoot}/dist/{projectRoot}'],
            },
          },
        },
      });

      builder.addNode({
        name: 'grandchild',
        type: 'lib',
        data: {
          root: 'libs/grandchild',
          targets: {
            build: {
              dependsOn: ['^build'],
              inputs: ['prod', 'deps'],
              executor: 'nx:run-commands',
              outputs: ['{workspaceRoot}/dist/{projectRoot}'],
            },
          },
        },
      });

      builder.addStaticDependency('parent', 'child', 'libs/parent/filea.ts');
      builder.addStaticDependency('child', 'grandchild', 'libs/child/fileb.ts');

      let projectGraph = builder.getUpdatedProjectGraph();
      let taskGraph = createTaskGraph(
        projectGraph,
        { build: ['^build'] },
        ['parent'],
        ['build'],
        undefined,
        {}
      );

      let nxJson = {
        namedInputs: {
          prod: ['!{projectRoot}/**/*.spec.ts'],
          deps: [{ dependentTasksOutputFiles: '**/*.d.ts', transitive: true }],
        },
        targetDefaults: {
          build: {
            dependsOn: ['^build'],
            inputs: ['prod', 'deps'],
            executor: 'nx:run-commands',
            options: {
              outputPath: 'dist/libs/{projectRoot}',
            },
            outputs: ['{options.outputPath}'],
          },
        },
      } as any;

      await tempFs.createFiles({
        'dist/libs/child/index.d.ts': '',
        'dist/libs/grandchild/index.d.ts': '',
      });

      const hasher = new InProcessTaskHasher(
        projectFileMap,
        allWorkspaceFiles,
        projectGraph,
        nxJson,
        null,
        {}
      );

      const transformed = transferProjectGraph(
        transformProjectGraphForRust(projectGraph)
      );
      const planner = new HashPlanner(nxJson, transformed);
      let plans = await assertHashPlan(
        taskGraph.tasks['parent:build'],
        taskGraph,
        hasher,
        planner
      );
      expect(plans).toMatchSnapshot();
    });
  });
});
