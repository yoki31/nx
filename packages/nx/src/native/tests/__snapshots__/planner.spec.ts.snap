// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task planner dependentTasksOutputFiles should depend on dependent tasks output files 1`] = `
{
  "parent:build": [
    "**/*.d.ts:dist/libs/child",
    "**/*.d.ts:dist/libs/grandchild",
    "AllExternalDependencies",
    "parent:!{projectRoot}/**/*.spec.ts",
    "parent:ProjectConfiguration",
    "parent:TsConfig",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/nx.json",
  ],
}
`;

exports[`task planner should be able to handle multiple filesets per project 1`] = `
{
  "parent:test": [
    "AllExternalDependencies",
    "child:!{projectRoot}/**/*.spec.ts",
    "child:ProjectConfiguration",
    "child:TsConfig",
    "env:MY_TEST_HASH_ENV",
    "parent:ProjectConfiguration",
    "parent:TsConfig",
    "parent:{projectRoot}/**/*",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/global1",
    "{workspaceRoot}/global2",
    "{workspaceRoot}/nx.json",
  ],
}
`;

exports[`task planner should build plans where the project graph has circular dependencies 1`] = `
{
  "child:build": [
    "AllExternalDependencies",
    "child:ProjectConfiguration",
    "child:TsConfig",
    "child:{projectRoot}/**/*",
    "parent:ProjectConfiguration",
    "parent:TsConfig",
    "parent:{projectRoot}/**/*",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/nx.json",
  ],
  "parent:build": [
    "AllExternalDependencies",
    "child:ProjectConfiguration",
    "child:TsConfig",
    "child:{projectRoot}/**/*",
    "parent:ProjectConfiguration",
    "parent:TsConfig",
    "parent:{projectRoot}/**/*",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/nx.json",
  ],
}
`;

exports[`task planner should hash executors 1`] = `
{
  "proj:lint": [
    "npm:@nx/devkit",
    "npm:@nx/eslint",
    "proj:ProjectConfiguration",
    "proj:TsConfig",
    "proj:{projectRoot}/**/*",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/global1",
    "{workspaceRoot}/nx.json",
  ],
}
`;

exports[`task planner should include npm projects 1`] = `
{
  "app:build": [
    "AllExternalDependencies",
    "app:ProjectConfiguration",
    "app:TsConfig",
    "app:{projectRoot}/**/*",
    "npm:react",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/nx.json",
  ],
}
`;

exports[`task planner should make a plan with multiple filesets of a project 1`] = `
{
  "parent:build": [
    "AllExternalDependencies",
    "parent:!{projectRoot}/**/*.spec.ts",
    "parent:ProjectConfiguration",
    "parent:TsConfig",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/nx.json",
  ],
  "parent:test": [
    "AllExternalDependencies",
    "parent:ProjectConfiguration",
    "parent:TsConfig",
    "parent:{projectRoot}/**/*",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/nx.json",
  ],
}
`;

exports[`task planner should plan non-default filesets 1`] = `
{
  "parent:build": [
    "AllExternalDependencies",
    "child:ProjectConfiguration",
    "child:TsConfig",
    "child:{projectRoot}/**/*",
    "parent:!{projectRoot}/**/*.spec.ts",
    "parent:ProjectConfiguration",
    "parent:TsConfig",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/nx.json",
  ],
}
`;

exports[`task planner should plan the task where the project has dependencies 1`] = `
{
  "parent:build": [
    "AllExternalDependencies",
    "child:ProjectConfiguration",
    "child:TsConfig",
    "child:{projectRoot}/**/*",
    "grandchild:ProjectConfiguration",
    "grandchild:TsConfig",
    "grandchild:{projectRoot}/**/*",
    "parent:ProjectConfiguration",
    "parent:TsConfig",
    "parent:{projectRoot}/**/*",
    "{workspaceRoot}/.gitignore",
    "{workspaceRoot}/.nxignore",
    "{workspaceRoot}/nx.json",
  ],
}
`;
